File: src/graph/matrix.ts
export type Input = [number, ...number[][]];
export type Output = number[][];

export function main(input: Input): Output {
  const length = input[0];

  const matrix: number[][] = new Array(length);
  for (let i = 0; i < length; i++) matrix[i] = new Array(length).fill(0);

  const list = input.filter((_, i): _ is number[] => i !== 0);

  for (const l of list) {
    const [vertex, , ...edges] = l;

    for (const e of edges) matrix[vertex - 1][e - 1] = 1;
  }

  return matrix;
}

File: src/graph/matrix.rs
type Output<'a> = &'a mut [&'a mut [i8]];

pub fn matrix<'a>(output: Output<'a>, vec: &[Vec<i8>]) -> Output<'a> {
    for v in vec.iter() {
        let vertex = v[0];
        let edges = &v[2..];

        for edge in edges.iter() {
            output[(vertex - 1) as usize][(edge - 1) as usize] = 1;
        }
    }
    output
}

#[cfg(test)]
mod tests {

    use super::*;

    #[test]
    fn should_generate_matrix() {
        let input = vec![vec![1, 2, 2, 4], vec![2, 1, 4], vec![3, 0], vec![4, 1, 3]];

        let mut output: [[i8; 4]; 4] = [[0; 4]; 4];
        let mut output_slices: Vec<&mut [i8]> =
            output.iter_mut().map(|row| row.as_mut_slice()).collect();

        assert_eq!(
            matrix(&mut output_slices, &input),
            [[0, 1, 0, 1], [0, 0, 0, 1], [0, 0, 0, 0], [0, 0, 1, 0],]
        );
    }
}
